### PR TITLE
speedup(~7x) of the clipping array inside scaling function 

### PR DESCRIPTION
**Pull request by ashish615**
- [ ] Closes #
- [ ] Tests included or not required because:

- [ ] Release notes not necessary because:

Hi,
We are submitting PR for speed up of the clipping part of scaling  function. 
|                 |  Time(sec)|
| -----------| ----- |
| Original | 11.82 |
| Updated | 1.59 |
| Speedup | 7.433962264 |

experiment setup : AWS r7i.24xlarge

```python
import time
import numpy as np

import pandas as pd

import scanpy as sc
from sklearn.cluster import KMeans

import os
import wget

import warnings



warnings.filterwarnings('ignore', 'Expected ')
warnings.simplefilter('ignore')
input_file = "./1M_brain_cells_10X.sparse.h5ad"

if not os.path.exists(input_file):
    print('Downloading import file...')
    wget.download('https://rapids-single-cell-examples.s3.us-east-2.amazonaws.com/1M_brain_cells_10X.sparse.h5ad',input_file)


# marker genes
MITO_GENE_PREFIX = "mt-" # Prefix for mitochondrial genes to regress out
markers = ["Stmn2", "Hes1", "Olig1"] # Marker genes for visualization

# filtering cells
min_genes_per_cell = 200 # Filter out cells with fewer genes than this expressed
max_genes_per_cell = 6000 # Filter out cells with more genes than this expressed

# filtering genes
min_cells_per_gene = 1 # Filter out genes expressed in fewer cells than this
n_top_genes = 4000 # Number of highly variable genes to retain

# PCA
n_components = 50 # Number of principal components to compute

# t-SNE
tsne_n_pcs = 20 # Number of principal components to use for t-SNE

# k-means
k = 35 # Number of clusters for k-means

# Gene ranking

ranking_n_top_genes = 50 # Number of differential genes to compute for each cluster

# Number of parallel jobs
sc._settings.ScanpyConfig.n_jobs = os.cpu_count()

start=time.time()
tr=time.time()
adata = sc.read(input_file)
adata.var_names_make_unique()
adata.shape
print("Total read time : %s" % (time.time()-tr))



tr=time.time()
# To reduce the number of cells:
USE_FIRST_N_CELLS = 1300000
adata = adata[0:USE_FIRST_N_CELLS]
adata.shape

sc.pp.filter_cells(adata, min_genes=min_genes_per_cell)
sc.pp.filter_cells(adata, max_genes=max_genes_per_cell)
sc.pp.filter_genes(adata, min_cells=min_cells_per_gene)
sc.pp.normalize_total(adata, target_sum=1e4)
print("Total filter and normalize time : %s" % (time.time()-tr))


tr=time.time()
sc.pp.log1p(adata)
print("Total log time : %s" % (time.time()-tr))


# Select highly variable genes
sc.pp.highly_variable_genes(adata, n_top_genes=n_top_genes, flavor = "cell_ranger")

# Retain marker gene expression
for marker in markers:
        adata.obs[marker + "_raw"] = adata.X[:, adata.var.index == marker].toarray().ravel()

# Filter matrix to only variable genes
adata = adata[:, adata.var.highly_variable]

ts=time.time()
#Regress out confounding factors (number of counts, mitochondrial gene expression)
mito_genes = adata.var_names.str.startswith(MITO_GENE_PREFIX)
n_counts = np.array(adata.X.sum(axis=1))
adata.obs['percent_mito'] = np.array(np.sum(adata[:, mito_genes].X, axis=1)) / n_counts
adata.obs['n_counts'] = n_counts


sc.pp.regress_out(adata, ['n_counts', 'percent_mito'])
print("Total regress out time : %s" % (time.time()-ts))

#scale

ts=time.time()
sc.pp.scale(adata,max_value=10)
print("Total scale time : %s" % (time.time()-ts))
```

https://github.com/scverse/scanpy/blob/706d4ef65e5d65e04b788831e7fd65dbe6b2a61f/scanpy/preprocessing/_scale.py#L202

### CODE REVIEW

**Comment by flying-sheep** [comment]
pre-commit.ci autofix

File: scanpy/preprocessing/_scale.py
Comment (Intron7) [inline scanpy/preprocessing/_scale.py]:
```suggestion
@numba.njit(parallel=True, cache = True)
```

Please cache the kernel.

File: scanpy/preprocessing/_scale.py
Comment (Intron7) [inline scanpy/preprocessing/_scale.py]:
``` suggestion
                if X[r,c] > a_max:
                     X[r,c] = a_max
                elif X[r,c] < a_min:
                     X[r,c] = a_min
```
This is even faster, because you safe some if statements for me it's 2X faster than your implementation.

**Review by Intron7 — CHANGES_REQUESTED** [review]
It's really a pity, that we can't just loop over all elements irrespective of C or F. But the performance looks really good in general. Please also check the pre-commit checks and remove the `tabs`

**Comment by flying-sheep** [comment]
Hi, I don’t have push permissions to IntelLabs/scanpy, so I can’t fix the merge conflicts. But you can easily do it yourself.

```bash
# make sure you are doing this from the command line, not e.g. VS Code
git fetch origin main:main
git merge main
git push
```

**Comment by Intron7** [comment]
Can you still please check the use of `tab`

**Comment by flying-sheep** [comment]
to make formatting easier, you should do `pre-commit install` like mentioned in the contributor guide: https://scanpy.readthedocs.io/en/stable/dev/getting-set-up.html#pre-commit

Please also fill out the checklist:

- if there is an issue closed by this, please link it
- you can check the box for tests, since this function already has tests. I added the `benchmark` label so we see that it actually makes things faster
- what’s missing is a release note entry: just edit `1.10.2.md` and add a line there please

**Review by Intron7 — APPROVED** [review]
Please add a release note

**Comment by flying-sheep** [comment]
Seems like we don’t have a benchmark for the clipping. Should we add one?

**Comment by Intron7** [comment]
I don't think it's necessary

File: docs/release-notes/1.10.2.md
Comment (Intron7) [inline docs/release-notes/1.10.2.md]:
```suggestion
* Speed up clipping of array in {func}`~scanpy.pp.scale` {pr}`3100` {smaller}`P Ashish & S Dicks`
```

**Review by Intron7 — CHANGES_REQUESTED** [review]
Please add me to the release note

**Comment by flying-sheep** [comment]
Hm, the warning making the docs build fail doesn’t seem to be related.

I’ll check if it is after all and merge this tomorrow

**Pull request merged**
